test_conformance/api/test_queue_properties_queries.cpp

```diff
@@ -0,0 +1,313 @@
+            context, deviceID, test_case.properties.data(), &error);
+        test_error(error, "clCreateCommandQueueWithProperties failed");
+    }
+    else
+    {
+        test_queue =
+            clCreateCommandQueueWithProperties(context, deviceID, NULL, &error);
+        test_error(error, "clCreateCommandQueueWithProperties failed");
+    }
+
+    std::vector<cl_queue_properties> check_properties;
+    size_t set_size = 0;
+
+    error = clGetCommandQueueInfo(test_queue, CL_QUEUE_PROPERTIES_ARRAY, 0,
+                                  NULL, &set_size);
+    test_error(error,
+               "clGetCommandQueueInfo failed asking for "
+               "CL_QUEUE_PROPERTIES_ARRAY size.");
+
+    if (set_size == 0 && test_case.properties.size() == 0)
+    {
+        return TEST_PASS;
+    }
+    if (set_size != test_case.properties.size() * sizeof(cl_queue_properties))
+    {
+        log_error("ERROR: CL_QUEUE_PROPERTIES size is %d, expected %d.\n",
+                  set_size,
+                  test_case.properties.size() * sizeof(cl_queue_properties));
+        return TEST_FAIL;
+    }
+
+    cl_uint number_of_props = set_size / sizeof(cl_queue_properties);
+    check_properties.resize(number_of_props);
+    error = clGetCommandQueueInfo(test_queue, CL_QUEUE_PROPERTIES_ARRAY,
+                                  set_size, check_properties.data(), NULL);
+    test_error(
+        error,
+        "clGetCommandQueueInfo failed asking for CL_QUEUE_PROPERTIES_ARRAY.");
+
+    if (check_properties.back() != 0)
+    {
+        log_error("ERROR: Incorrect last property value - should be 0!\n");
+        return TEST_FAIL;
+    }
+
+    check_properties.pop_back();
+    test_case.properties.pop_back();
+
+    if (check_properties != test_case.properties)
+    {
+        for (cl_uint i = 0; i < test_case.properties.size(); i = i + 2)
+        {
+            cl_queue_properties set_property = test_case.properties[i];
+            cl_queue_properties set_property_value =
+                test_case.properties[i + 1];
+
+            std::vector<cl_mem_properties>::iterator it = std::find(
+                check_properties.begin(), check_properties.end(), set_property);
+
+            if (it == check_properties.end())
+            {
+                log_error("ERROR: Property name not found ... 0x%x\n",
+                          set_property);
+                return TEST_FAIL;
+            }
+            else
+            {
+                if (set_property_value != *std::next(it))
+                {
+                    log_error("ERROR: Incorrect property value expected 0x%x, "
+                              "obtained 0x%x\n",
+                              set_property_value, *std::next(it));
+                    return TEST_FAIL;
+                }
+            }
+        }
+        log_error("ERROR: ALL properties and values matched but order "
+                  "incorrect!\n");
+        return TEST_FAIL;
+    }
```

Comment:
That property comparison code is the same across all three tests. We should write a common utility function using the new `cl_properties` type.

Reply:
Fixed.